File: 8/Java/Solution.java
public class Solution {

    public static int sum;

    public int nQueens(int n) {
        sum = 0;
        int[] cols = new int[n];
        helper(cols, n, 0);
        return sum;
    }

    private void helper(int[] cols, int n, int row) {
        if (row == n) {
            sum++;
            return;
        }

        for (int i = 0; i < n; i++) {
            if (isValid(cols, row, i)) {
                cols[row] = i;
                helper(cols, n, row + 1);
            }
        }
    }

    private boolean isValid(int[] cols, int row, int col) {
        for (int i = 0; i < row; i++) {
            if (cols[i] == col) {
                return false;
            }
            
            if ((row - i) == Math.abs(col - cols[i])) {
                return false;
            }
        }
        return true;
    }

    public static void main(String[] args) {

        System.out.println(new Solution().nQueens(8));
    }
}

File: 8/Python/Solution.py
# -*- coding:utf-8 -*-


class Solution:
    def nQueens(self, n):
        result = [0] * n
        results = []
        n_row = 0
        while n_row < n:
            pos = self.findLegalPosition(result, n_row, n)
            # print pos
            if pos is not None:
                result[n_row] = pos
                if n_row == n - 1:
                    results.append(result[:])

            if pos is None or n_row == n - 1:
                while 1:
                    n_row -= 1
                    if n_row < 0:
                        #print results
                        return len(results)

                    pos = self.findLegalPosition(
                        result, n_row, n, back_tracing=True)
                    if pos:
                        result[n_row] = pos
                        break
            n_row += 1

    def findLegalPosition(self, arr, n_row, n, back_tracing=False):
        start = 0
        if back_tracing:
            start = arr[n_row] + 1
        for col in range(start, n):
            ok = True
            for row in range(n_row):
                if arr[row] == col:
                    ok = False
                    break
                if n_row - row == abs(col - arr[row]):
                    ok = False
                    break
            if ok:
                return col
        return None


print(Solution().nQueens(8))
